experimental/tls_attestation/src/certificate.rs

```diff
@@ -0,0 +1,121 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use oak_attestation_common::{get_sha256, Report};
+use openssl::{
+    asn1::Asn1Time,
+    bn::{BigNum, MsbOption},
+    hash::MessageDigest,
+    pkey::{PKey, Private},
+    rsa::Rsa,
+    x509::{
+        extension::{BasicConstraints, KeyUsage, SubjectAlternativeName, SubjectKeyIdentifier},
+        X509NameBuilder, X509,
+    },
+};
+
+// X.509 certificate parameters.
+// https://tools.ietf.org/html/rfc5280
+const RSA_KEY_SIZE: u32 = 2048;
+// Version is zero-indexed, so the value of `2` corresponds to the version `3`.
+const CERTIFICATE_VERSION: i32 = 2;
+// Length of the randomly generated X.509 certificate serial number (which is 20 bytes) excluding
+// the most significant bit.
+const SERIAL_NUMBER_SIZE: i32 = 159;
+const CERTIFICATE_EXPIRATION_INTERVAL_IN_DAYS: u32 = 1;
+// TODO(#1867): Add remote attestation support.
+const TEST_TEE_MEASUREMENT: &str = "Test TEE measurement";
+
+/// Placeholder function for collecting TEE measurement of remotely attested TEEs.
+fn get_example_tee_report(public_key: &[u8]) -> Report {
+    let public_key_hash = get_sha256(public_key);
+    Report::new(
+        TEST_TEE_MEASUREMENT.to_string().as_bytes(),
+        &public_key_hash,
+    )
+}
+
+/// Convenience structure for creating X.509 certificates.
+/// https://tools.ietf.org/html/rfc5280
+pub struct CertificateAuthority {
+    pub key_pair: PKey<Private>,
+    pub certificate: X509,
+}
+
+pub struct Identity {
+    pub private_key: Vec<u8>,
+    pub certificate: Vec<u8>,
+}
+
+impl CertificateAuthority {
+    /// Creates a self-signed X.509 certificate and a corresponding private key.
+    /// The certificate is bound to `domain_name` and contains TEE credentials in a custom X.509
+    /// extension.
+    pub fn generate_certificate(domain_name: &str) -> anyhow::Result<Identity> {
```

Comment:
Domain name seems irrelevant (in fact misleading) since it's always going to be a self signed cert, right?